lib/active_set/sorting/operation.rb

```diff
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+require_relative '../attribute_instruction'
+
+class ActiveSet
+  module Sorting
+    class Operation
+      def initialize(set, instructions_hash)
+        @set = set
+        @instructions_hash = instructions_hash
+      end
+
+      def execute
+        attribute_instructions = @instructions_hash
+                                   .flatten_keys
+                                   .map { |k, v| AttributeInstruction.new(k, v) }
+
+        activerecord_strategy = ActiveRecordStrategy.new(@set, attribute_instructions)
+        if activerecord_strategy.executable_instructions == attribute_instructions
+          activerecord_sorted_set = activerecord_strategy.execute
+        end
+
+        return activerecord_sorted_set if attribute_instructions.all?(&:processed?)
+
+        EnumerableStrategy.new(@set, attribute_instructions).execute
+      end
+
+      def operation_instructions
+        @instructions_hash.symbolize_keys
+      end
+    end
+
+    class EnumerableStrategy
+      def initialize(set, attribute_instructions)
+        @set = set
+        @attribute_instructions = attribute_instructions
+      end
+
+      def execute
+        @set.sort_by do |item|
+          @attribute_instructions.map do |instruction|
+            sortable_numeric_for(instruction, item) * direction_multiplier(instruction.value)
+          end
+        end
+      end
+
+      def sortable_numeric_for(instruction, item)
+        value = instruction.value_for(item: item)
+
+        if value == true
+          return 1
+        elsif value == false
+          return 0
+        elsif value.is_a?(Numeric)
+          return value
+        elsif value.is_a?(String) || value.is_a?(Symbol)
```

Comment:
Strings and Symbols are the trickiest of the bunch. In the comment below I annotate how a value is transformed step-by-step. The key points are that we need

1. the value to have a fixed integer length (so that strings of different lengths are properly sorted; if we converted simply to an integer, longer strings regardless of starting letter would be sorting according to their length)
2. the value to be able to handle large strings (thus `to_r` which returns a `Rational` instead of `to_f` which returns a `Float` as floats would truncate values after a certain point)
3. the value will be sorted is "ascii-etical" order not alphabetical order

Reply:
This doesn't necessarily have to be done as part of this PR, but I'd like to see some unit tests for this method. I think there's a lot of tricky stuff going on in here so it would be really helpful to have those tests because (a) they provide some documentation and (b) they'll help us prevent regressions in the future.

I know there are some higher level specs that might help us catch those regressions, but having specs on this method specifically will make troubleshooting those things easier if we catch them.

Reply:
I added some basic unit tests and extracted this method out as a helper.